lib/Puppeteer.js

```diff
@@ -18,6 +18,8 @@ const BrowserFetcher = require('./BrowserFetcher');
 const Errors = require('./Errors');
 const DeviceDescriptors = require('./DeviceDescriptors');
 
+const PRODUCT = process.env.PUPPETEER_PRODUCT || process.env.npm_config_puppeteer_product || process.env.npm_package_config_puppeteer_product;
```

Comment:
Yes, there may be a better place for this.

The main motivation here is to maintain some sort of privileged state about which browser we're talking to. This is both for Puppeteer API consumers and the Puppeteer implementation itself, since both internally and externally some browser-specific code may be needed from time to time.

The same holds true even in the case of `puppeteer-core`: the library doesn't download any browser binary for us, but which browser is it ultimately talking to? Or am I mistaken?

Given that, do you have any suggestions?

Reply:
^ @mathiasbynens 

Mentioning you in case this question got lost in a sea of notifications. :)

Reply:
The problem is that `puppeteer-core` might be run outside of Node.js, where `process.*` is not available. In the Launcher, we wrap any usage of `process.env` in `if (!this._isPuppeteerCore) { ... }` for that reason.

Reply:
Maybe I'm misunderstanding. What problems do you see with keeping the logic local to `lib/Launcher.js`?

Reply:
I think I've addressed both our concerns with the latest change. :) The `PUPPETEER_PRODUCT` stuff is only used if we're not running `puppeteer-core`, and I changed Puppeteer to lazily load an appropriate Launcher based on a `product` option, if available, falling back on environment variables.